### ISSUE TITLE
SM & RStudio autoStop does not work if user does not click on button 'close the session' before closing its web browser

### ISSUE THREAD

**Issue by Ludo171** [awslabs/service-workbench-on-aws#185]
#### Description :
Related to the new feature proposed by PR #171 (commit 1e8b067f96c463a4b3b27f8b625d2f2f1591d452). We expect that this new feature would turn RStudio workspaces to sleep after 1hour of inactivity, and Sagemaker Notebooks after a custom idletime specified in the workspace configuration. I used the Sagemaker Notebook and RStudio configurations coming out-of-the-box from 81c60ef537cb45b55dc701b7d0a0dfbfbb26c81c.

The behavior works as expected if I do not try to connect on the workspaces at all. In contrast, if I connect on the workspaces and execute a few instructions before closing the window, I realized that they were still up and running 5 hours afterwards.

#### Reproduce the bug :
- Create an RStudio workspace
- Connect onto it and execute a few instructions ('variable <- 4 +4') suffices.
- Close the window, and your web browser.
- Connect on Service Workbench more than 60min afterwards
--> Assess that the workspace is still up and running
--> On the AWS console, assess that the EC2 instance is still up and running
--> The same behavior goes for Sagemaker notebooks if you specify 60 min of AutoStopIdleTime in the configurations.

On the opposite, if I do not perform step 2 (connect and execute instructions), then the workspace automatically stops after 1 hour.

**Comment by SanketD92** [comment]
1. What type of instructions did you run on the workspaces? Would they be expected to run in the background on the instances?
2. Could you confirm if you "Quit" the RStudio session before exiting the portal?

FYI - As for the Sagemaker auto-stop feature, we're using this [script](https://raw.githubusercontent.com/aws-samples/amazon-sagemaker-notebook-instance-lifecycle-config-samples/master/scripts/auto-stop-idle/autostop.py) to check for its idle state from the last activity.

**Comment by Ludo171** [comment]
Hi @SanketD92 

1) Here are the very basic instructions that I ran on RStudio and Sagemaker :
![CaptureRstudio](https://user-images.githubusercontent.com/33128425/96127529-b8319b80-0ef4-11eb-9143-a15a9ae85d25.PNG)
![CaptureSagemakr](https://user-images.githubusercontent.com/33128425/96127554-b9fb5f00-0ef4-11eb-803f-67ea869f3977.PNG)  

2) I did not click on any 'Logout' button before closing my web browser for both Sagemaker and RStudio. I did it on purpose to mimik the behavior of potential researchers and students working on workspaces created by Service Workbench. 

I reproduced the test again to probe your point : I performed the same steps as described in `Reproduce the bug` , but this time I clicked on the 'Close session' and 'Logout' button before closing my window and web browser. And it worked : both workspaces Sagemaker and RStudio stopped after 1 hour.

In conclusion I think that the behavior is encouraging but not yet exactly what would be needed on the true field : it seems that the sessions are not properly closed when some user closes his/her web browser, which can result in some workspaces running for a while for no purpose : runaway costs... It makes the auto-stop feature unreliable until we find a fix to handle such use-cases.

Did you manage to reproduce the same behavior ?
Do you have more details about the way sessions work on RStudio and Sagemaker ? Could we combine other mechanisms to make sure that no useless session is kept alive and prevents the auto-stop to work as expected ?

**Comment by SanketD92** [comment]
Hi @Ludo171 

The SageMaker workspaces get auto-stopped when their internal CPU resource consumption is at a minimum consistently for the time period configured for auto-stop. 

As for RStudio, the `check-idle.py` script checks the number of active sessions and calculates uptime from the last active session time. (Also, the safest way for RStudio to end a session is to quit the R session button so as to prevent unintended data loss as mentioned in instructions [here](https://github.com/awslabs/service-workbench-on-aws/tree/master/main/solution/machine-images))

I tested the scenario you mentioned for SageMaker (i.e connecting to it); and the auto-stop feature still worked, without the need to hit the Logout button on that instance. 
For RStudio, I'm planning to conduct a test with a shorter time period for auto-stop. Stay tuned for that.

FYI - According to the commit id you've mentioned, it seems you're still a few commits behind the latest release [v1.3.0](https://github.com/awslabs/service-workbench-on-aws/releases/tag/v1.3.0)

**Comment by SanketD92** [comment]
@Ludo171,

RStudio auto-stop logic update has been merged to the mainline branch. Please feel free to redeploy the machine-images SDC and use the updated RStudio AMI to test the functionality.

As for the SageMaker auto-stop issue you reported, I wasn't able to reproduce that with your set of instructions. The logic for SageMaker auto-stop has always been derived from the code published [here](https://raw.githubusercontent.com/aws-samples/amazon-sagemaker-notebook-instance-lifecycle-config-samples/master/scripts/auto-stop-idle/autostop.py), so there _shouldn't_ be any change of behavior between built-in and SC. Let us know if you're still seeing that.

**Comment by nguyen102** [comment]
@SanketD92 can this ticket be closed now that the PR has been merged?

**Comment by Ludo171** [comment]
I made some user tests on a version of Service Workbench that includes PR #188 changes.
And it worked !
RStudio workspace stopped automatically even after using it for a few instructions.

As for Sagemaker notebooks, I figured that the auto-stop feature works, but takes more time than specified. I set the parameter to 15 min in my test configurations and it took between 45min and 1hour today to stop after I closed my web browser. I should have been more patient during my first tests.

In my opinion this issue can be closed.

**Comment by SanketD92** [comment]
Thanks for confirming, @Ludo171.

@nguyen102, yes. I'll close it now. Thanks for following through on this.
